llvm/lib/Target/AMDGPU/SIISelLowering.cpp

```diff
@@ -1604,6 +1604,14 @@ bool SITargetLowering::isLegalAddressingMode(const DataLayout &DL,
         return false;
     }
 
+    if (AS == AMDGPUAS::CONSTANT_ADDRESS && AM.BaseOffs < 0) {
```

Comment:
should also cover the constant_32bit (but again, this whole thing is just a weak heuristic for might-use-scalar-loads)

Reply:
I specifically didn't want to include constant_32bit because I saw a case where LLPC was using that address space for an s_buffer_load, not an s_load. But I did not look into _why_ that was happening. I have never understood constant_32bit.

Reply:
It's a CONSTANT_ADDRESS truncated to the low 32-bits, where the high bits are assumed a constant from an attribute. The addressing modes should behave the same as constant, it's just implicitly promoted to the 64-bit pointer when codegenned

Reply:
> should also cover the constant_32bit

Done